OrchardCore.Commerce/Views/ShoppingCart/Index.cshtml

```diff
@@ -21,30 +20,36 @@
             <div class="col-xs-12 col-sm-1">@T["Price"]</div>
             <div class="col-xs-12 col-sm-1">@T["Action"]</div>
         </div>
-        @{
-            int line = 0;
-        }
-        @foreach (ShoppingCartLineViewModel item in Model.Lines)
+        @for (var line = 0; line < Model.Lines.Count; line++)
         {
+            var item = Model.Lines[line];
+            var name = $"cart.lines[{line}]";
+            var attributes = item
+                .Attributes
+                .Select((attribute, index) => (attribute.Value, Type: attribute.GetType().Name, Index: index))
+                .Where(pair => pair.Value.UntypedValue != null)
+                .ToList();
+
             <div class="row border-bottom mt-3 pb-2">
                 <div class="col-xs-12 col-sm-1">
-                    @{
-                        int i = 0;
-                        var name = $"cart.lines[{line++}]";
-                    }
                     <input form="shopping-cart-update" type="number" min="0" name="@(name).Quantity" value="@item.Quantity" class="form-control" />
                     <input form="shopping-cart-update" type="hidden" name="@(name).ProductSku" value="@item.ProductSku" />
-                    @foreach (IProductAttributeValue attr in item.Attributes.Select(attr => attr.Value).Where(val => val.UntypedValue != null))
+                    @foreach (var (attribute, type, index) in attributes)
                     {
-                        <shape Type="@attr.GetType().Name" alternate="@(attr.GetType().Name)_hidden" form="shopping-cart-update" prop-index="@(i++)" prop-attributeValue="@attr" prop-namePrefix="@name" />
+                        <shape Type="@(type)"
+                               alternate="@(type)_hidden"
+                               form="shopping-cart-update"
+                               prop-index="@index"
+                               prop-attributeValue="@(attribute)"
+                               prop-namePrefix="@name" />
                     }
                 </div>
                 <div class="col-xs-12 col-sm-9">
                     <a href="@item.ProductUrl">@item.ProductName</a>
                     @if (item.Attributes != null)
                     {
                         <ul class="cart-product-attributes list-inline">
-                            @foreach (IProductAttributeValue attr in item.Attributes.Select(attr => attr.Value).Where(val => val.UntypedValue != null))
+                            @foreach (var attr in item.Attributes.Select(attr => attr.Value).Where(val => val.UntypedValue != null))
                             {
                                 <shape tag="li" Type="@attr.GetType().Name" prop-attribute="@attr" prop-label="@attr.Label" />
                             }
```

Comment:
This can now become:

```suggestion
                            @foreach (var (attribute, type, _) in attributes)
                            {
                                <shape tag="li" Type="@(type)" prop-attribute="@(attribute)" prop-label="@(attribute).Label" />
                            }
```